Setup: make `update` change config and implement `status`

klees opened the pull request:
Dear colleagues,

this implements two feature requests for the setup: [Add `status` Command](https://docu.ilias.de/goto.php?target=wiki_1357_Setup_-_Add_status_Command) and [Let `update` Command Change Configs](https://docu.ilias.de/goto.php?target=wiki_1357_Setup_-_Let_update-Command_change_configs) and also cleans up some things a little. Although we advice to create small PRs I hope it is easier for every individual maintainer to look through changes in their components in bulk. Changes regarding components of other maintainers should be mostly repetitive and contain little logic.

This is what happened in general:

* I removed `Agent::getConfigInput`. When building the setup initially I though we would need this for the webbased GUI, but this will be removed soon. Anyway, it was never really implemented anywhere.
* I introduced [some infrastructure for metrics](https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/src/Setup/Metrics) before that is put to use now to implement the status command. `Agent`s can now provide an `Objective` for the status command via `getStatusObjective`. That `Objective` is supposed to collect metrics and store them to a `Storage` for metrics. I have provided a base class that should make it fairly easy and obvious how to implement an appropriate `Objective`. This was exercised for various components, mostly to collect the current configuration from the installation.
* I decoupled two widely used `Objective`s from configs provided by agents to make them more easily usable. These are `ilIniFilesLoadedObjective` and `ilDatabaseInitializedObjective`. These lead to some simplifications in various `Objective`s.
* This also made it possible to remove the hard dependency on a configuration for the `update` command (or `Agent::getUpdateObjective`, respectively). Although the signature for `Agent::getUpdateObjective` did not change, `Agent`s now need to update their config when one is provided (and do other stuff even without config if necessary on update). I implemented that for all components where this is required. To make this work I needed to drop the dependency on the config in various places.
* This in turn made it possible to make the `config` option for the `update` command optional.

I will go into a little bit more detail for some of the involved maintainers.

I will improve on two points in this PR, which is documentation for the metrics and sorting the output for `status` to separate configs and other metrics. I still already open this PR to give involved maintainers a chance to have a look already.

As always, feel free to ask questions, give feedback or request changes. And hopefully, click "Approve", of course.

Best regards!

klees committed:
Setup: removed Agent::getConfigInput

This method initially was meant to be used in a WebGUI for the setup. This
won't be implemented for a foreseeable future and, if it will be implemented,
won't need an elaborate mechanism to input configuration for the configs.
The method was never really implemented or used, so can be removed without
any further ado.

klees committed:
Setup/Mail: minor cleanup

klees committed:
Setup: make `update` change config

Currently, update only performs database updates but no changes of configuration.
On the one hand, this is irritating users of the web setup. On the other hand,
this will become a requirement if we remove the websetup, because we would lose
the possibility to change (some) config, then.

klees committed:
Setup: let agents have an objective to for status command

Added Setup\Agent::getStatusObjective and according stubs in the implementors
of that interface.

renaming

klees committed:
Setup: add minimum version of `status` command

klees committed:
Setup: fix return type in Metric

klees committed:
Setup: config-parameter to getUpdateObjective is optional

klees committed:
Setup: config option to update command is optional

klees committed:
Setup: factor out a base objective for metric collection

klees committed:
Setup: implemented metric collection for various components

klees commented:
Dear @alex40724,

please have a look into this PR, I made changes in these components of yours:

* Modules/LearningModule: some minor cleanup and adjustments
* Service/UICore: same here
* Modules/SystemFolder: some minor cleanup, but mostly adding the objective for the metrics and making update actually update the config
* Service/MediaObject: same here
* Services/Administration: use simplified database objective

Would you also have a look into Services/Utilities? I did the same here than in Modules/SystemFolder and the MediaObject.

Best regards!

klees commented:
Dear @mjansenDatabay,

please have a look into this PR, I made changes in these components of yours:

* Modules/Forum: some minor cleanup and adjustments
* Services/Mail: same here

Best regards!

klees commented:
Dear @mstuder,

please have a look into this PR, I made changes in these components of yours:

* Services/Component: some minor cleanup and adjustments

Best regards!

klees commented:
Dear @fneumann,

please have a look into this PR, I made changes in these components of yours:

* Services/Language: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config
* Services/MathJax: same here

Best regards!

klees commented:
Dear @gvollbach,

please have a look into this PR, I made changes in these components of yours:

* Services/PDFGeneration: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config

Best regards!

klees commented:
Dear @rschenk0,

please have a look into this PR, I made changes in these components of yours:

* Services/VirusScanner: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config

Best regards!

klees commented:
Dear @Amstutz,

please have a look into this PR, I made changes in these components of yours:

* Services/Style: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config

Best regards!

klees commented:
Dear @smeyer-ilias,

please have a look into this PR, I made changes in these components of yours:

* Services/Logging: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config
* Services/User: some minor adjustments

Best regards!

alex40724 reviewed (changes requested):
Hi @klees,

looks good to me. One issue: I think we should mention the collection of the administrators personal data (name, email) on https://docu.ilias.de/goto_docu_wiki_wpage_6374_1357.html under the privacy section. If monitoring systems will start to collect these information it might even leave the server and be transferred to other systems (yes, this is not done by ILIAS, but I think everyone should be aware that this data is included).

klees commented:
Dear @chfsx,

you somehow seem to attract components that are involved in the setup =)

I made various changes, please have a look into the PR:

* Services/GlobalScree: some minor cleanup and adjustments
* Services/BackgroundTasks: some minor cleanup and adjustments, but mostly adding the objective for the metrics and making update actually update the config
* Services/GlobalCache: same here
* Services/Http: same here
* Services/Preview: same here
* Services/FileSystem: almost the same here, but since we cannot simply move directories by changing them in the config, I also added `ilFileSystemConfigNotChangedObjective` that simply denies according config changes.
* Services/Database: as for the other services, but: I also added a `ilDatabaseInitializedObjective` that only relies on the ini files to make using the database in the setup easier. Also, there is a lot of cruft in the database classes caused by the heavy usage of globals and constants. It might be worthwhile to clean up and refactor here some time. 

As always, feel free to contact me via skype or similar, if you want to discuss stuff. Since the changes in your components are not that straight forward and involve some decisions, this might be easier then discussing textually here. This of course also goes for everybody else!

Best regards!

klees commented:
Thx @alex40724, nice catch! I adjusted the privacy section accordingly.

mjansenDatabay reviewed (approved):
LGTM!

klees committed:
Setup: extract config metrics for output

chfsx reviewed (approved):
Thanks @klees for all this work, I had a look at my parts and they're looking absolutely fine!

Services/Http: This is an old service maintained by @mjansenDatabay  as far as I know, but LGTM as well
Services/FileSystem: Thanks for ilFileSystemConfigNotChangedObjective, this is definitely needed ATM
Services/Database: And yes, DB need a lot of work in the next time...

chfsx commented:
@klees and in addition for the Services/Component-changes: LGTM as well!

mjansenDatabay commented:
@chfsx / @klees : AFAIK I do not maintain `Services/Http` (yes, the `ilProxySettings` is mine), but `Services/Html`.

chfsx commented:
@mjansenDatabay OC, sorry :-) I opened the wrong directory... In Services/Http @smeyer-ilias is listed as implicit maintainer

rschenk0 reviewed (approved):
Guido reviewed and says OK.

klees committed:
Setup: more docu for `status` command and facilities

smeyer-ilias reviewed (approved):
Thanks, looks good to me.

Amstutz reviewed (approved):
Thx a lot for this large refactoring. The changes in System Styles are good for me. Thx!

fneumann commented:
Dear @rklees, it looks all fine for me to. Feel free to commit setup related changes that do not directly change the code processing language files or database entries.
